utils/signed_files.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include <stdlib.h>
+
+#include "common/curl-p/trit.h"
+#include "common/kerl/kerl.h"
+#include "common/model/bundle.h"
+#include "common/sign/normalize.h"
+#include "common/sign/v1/iss_curl.h"
+#include "common/trinary/trit_long.h"
+#include "common/trinary/tryte_ascii.h"
+#include "utils/merkle.h"
+#include "utils/signed_files.h"
+
+/*
+ * Private functions
+ */
+
+static retcode_t validate_signature(char const *const signature_filename,
+                                    tryte_t const *const public_key,
+                                    size_t depth, size_t index,
+                                    flex_trit_t *const digest,
+                                    bool *const valid) {
+  retcode_t ret = RC_OK;
+  Curl curl;
+  FILE *fp = NULL;
+  ssize_t read = 0;
+  char *line = NULL;
+  size_t len = 0;
+  byte_t normalized_digest[HASH_LENGTH_TRYTE];
+  trit_t normalized_digest_trits[HASH_LENGTH_TRIT];
+  trit_t sig_digests[3 * HASH_LENGTH_TRIT];
+  trit_t sig_trits[NORMALIZED_FRAGMENT_LENGTH * HASH_LENGTH_TRYTE * RADIX];
+  trit_t root[HASH_LENGTH_TRIT];
+  trit_t public_key_trits[HASH_LENGTH_TRIT];
+  trit_t siblings[depth * HASH_LENGTH_TRYTE * RADIX];
+
+  if ((fp = fopen(signature_filename, "r")) == NULL) {
+    return RC_UTILS_FAILED_TO_OPEN_FILE;
+  }
+
+  curl.type = CURL_P_81;
+  init_curl(&curl);
+
+  normalize_hash(digest, normalized_digest);
+  for (int c = 0; c < HASH_LENGTH_TRYTE; ++c) {
+    long_to_trits(normalized_digest[c], &normalized_digest_trits[c * RADIX]);
+  }
+
+  int i;
+  for (i = 0; i < 3 && (read = getline(&line, &len, fp)) > 0; i++) {
+    line[--read] = '\0';
+    if (read != NORMALIZED_FRAGMENT_LENGTH * HASH_LENGTH_TRYTE) {
+      ret = RC_UTILS_INVALID_SIG_FILE;
+      goto done;
+    }
+    trytes_to_trits((tryte_t *)line, sig_trits, read);
+    iss_curl_sig_digest(
+        sig_digests + i * HASH_LENGTH_TRIT,
+        normalized_digest_trits + i * NORMALIZED_FRAGMENT_LENGTH * RADIX,
+        sig_trits, RADIX * read, &curl);
+    curl_reset(&curl);
+  }
+  if (i != 3 || read < 0) {
+    ret = RC_UTILS_INVALID_SIG_FILE;
+    goto done;
+  }
+
+  iss_curl_address(sig_digests, root, sizeof(sig_digests), &curl);
+  if ((read = getline(&line, &len, fp)) > 0) {
```

Comment:
Memleak, you should free this. `This buffer should be freed by the user program even if getline() failed.`

Reply:
Ditto

Reply:
Sorry, it was free'd in the other function, but not this one, fixing it rn